scripts/stats.js

```diff
@@ -0,0 +1,142 @@
+// Description
+//   Collects and yields general slack statistics for analysis and insight
+//
+// Commands:
+//   `!stats (rooms|commands)` - Yields general slack statistics for analysis and insight
+
+// Default stats upon reset
+DEFAULT_STATS = {rooms: {}, commands: {}};
+
+/////////////////////
+// HELPER COMMANDS //
+/////////////////////
+
+// Increment counter in map, first setting to 0 if it does not exist
+function incrementCounter(map, entry) {
+    if (!(entry in map)) {
+        map[entry] = 0;
+    }
+    map[entry]++;
+}
+
+// Returns a sorted list of object entries
+function getSortedEntries(object) {
+    var entries = [];
+    for (var entry in object) entries.push([entry, object[entry]])
+    return entries.sort((a, b) => b[1] - a[1]);
+}
+
+// Retrieves stored slack stats, setting them to default values if they do not exist
+function getStats(robot) {
+    var stats = robot.brain.get('stats');
+    if (!stats) {
+        robot.brain.set('stats', DEFAULT_STATS);
+        stats = robot.brain.get('stats');
+    }
+    return stats;
+}
+
+//////////////////////
+// HANDLER COMMANDS //
+//////////////////////
+
+// Handles room stat
+function handleRoomStat(stats, res) {
+    // If room was not a public channel and not a private channel, exit
+    var room = res.message.room;
+    if (room[0] != 'C' && room[0] != 'G') return;
```

Comment:
Could you explain what this line does? I'm a little confused.

Reply:
room ids have letter prefixes which denote whether they are of type Channel (aka public channels, 'C'), Group (aka private channels, 'G') or Direct (aka personal messages, prefix 'D'). I check this first character to basically check the room being passed is a valid public/private channel

Reply:
Are you sure we want private room statistics being included?

Reply:
yeah that's a valid question. I came to the conclusion it was fine as it's only tracking message count, which seems pretty non-intrusive to me

Reply:
If people have a problem with it though, happy to just limit it to public channels

Reply:
In fact I think that might be a good idea. Private channels aren't really of interest as a statistic anyway beyond the total number of private channel messages, which is a statistic we're already given

Reply:
Removed in latest commit

Reply:
Yeah, there's also the fact that some private channels would like to remain a secret

Reply:
ah good point I thought they still showed up in list of channels